packages/tangle/message.go

```diff
@@ -290,20 +273,101 @@ func NewMessage(strongParents, weakParents, dislikeParents, likeParents MessageI
 			ParentsCount: uint8(likeParentsCount),
 			References:   sortedLikeParents,
 		})
-		parentsBlocksCount++
+	}
+
+	return newMessageWithValidation(MessageVersion, parentsBlocks, issuingTime, issuerPublicKey, msgPayload, nonce, signature, sequenceNumber)
+}
+
+/**
+newMessageWithValidation creates a new message while performing ths following syntactical checks:
+1. A Strong Parents Block must exist.
+2. Parents Block types cannot repeat.
+3. Parent count per block 1 <= x <= 8.
+4. Parents unique within block.
+5. Parents lexicographically sorted within block.
+6. ParentsCount within each block must match the parents actually parsed within the block.
+7. A Parent(s) repetition is only allowed when it occurs across Strong and Like parents
+8. Blocks should be ordered by type in ascending order
+**/
+func newMessageWithValidation(version uint8, parentsBlocks []ParentsBlock, issuingTime time.Time,
+	issuerPublicKey ed25519.PublicKey, msgPayload payload.Payload, nonce uint64,
+	signature ed25519.Signature, sequenceNumber uint64) (result *Message, err error) {
+	// Validate strong parent block
+	if len(parentsBlocks) == 0 || parentsBlocks[StrongParentType].ParentsType != StrongParentType ||
```

Comment:
Unlike the other constructor `NewMessage`, this constructor doesn't sort or changes the contents of the inputs. It just checks their validity. If for example a deserialized message doesn't have StrongParentType as the first block it is invalid.

So I don't think I am making any assumptions here. I am simply checking if indeed we have a strong type and it is the first block.

Reply:
You are not necessarily checking that is the first block: you are assuming that `StrongParentType` will always have index 0; that's why I find that second condition strange to read: it seems you are using `parentBlocks` as a map.
Since on line 316 you are checking for a count mismatch you should also check the parentBlockCount and parentsBlock index matches the ParentsType of the block.

Reply:
I can change StrongParentType to 0..
The linter might give me a magic number error but I can ignore it (which is why I didn't do so in the first place)..

I didn't understand what you wanted me to do on line 316 (now 318)..
`parentBlockCount` is not a parameter of the constructor. The deserializer uses it for parsing so it can't have a mismatch.
I don't think the parentBlock index must match the block type..